### PR TITLE
Create a core/_flow.py of middle layer of protected flows.

### PR DESCRIPTION
**Pull request by crazygao**
# Description

Create a layer

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 1]
```diff
@@ -2,232 +2,29 @@
...
-from promptflow._utils.yaml_utils import load_yaml, load_yaml_string
+from promptflow._utils.yaml_utils import load_yaml
 from promptflow.exceptions import ErrorTarget, UserErrorException
 
 logger = get_cli_sdk_logger()
 
+Flow = FlowCore
+FlowBase = FlowBaseCore
+FlowContext = FlowContextCore
+AsyncFlow = AsyncFlowCore
```
Comment (crazygao) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:24]: Change this for not changing too much original imports.

File: src/promptflow/promptflow/_core/_flow.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+import abc
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/_flow.py:5]: The file should be in core/ not _core/

File: src/promptflow/promptflow/_core/_flow.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        :type data: dict
+        :param params_override: Parameters to override, defaults to None
+        :type params_override: typing.Optional[list]
+        :return: Class to use for deserializing the data & its "type". Type will be None if no override is provided.
+        :rtype: tuple[class, typing.Optional[str]]
+        """
+        return cls, "flow"
+
+
+class Flow(FlowBase):
```
[the remark this answers is not in the document]
Comment (crazygao) [reply src/promptflow/promptflow/_core/_flow.py:140]: Yep

File: src/promptflow/promptflow/_constants.py [hunk 4]
```diff
@@ -22,6 +22,8 @@
 LANGUAGE_KEY = "language"
 USER_AGENT_OVERRIDE_KEY = "user_agent_override"
 
+DAG_FILE_NAME = "flow.dag.yaml"
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_constants.py:25]: Prefer name this as : DEFAUT_FLOW_YAML_FILE_NAME

File: src/promptflow/promptflow/core/_errors.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+from promptflow.exceptions import ErrorTarget, SystemErrorException, UserErrorException
+
+
+class CoreError(UserErrorException):
+    """SDK base class, target default is CONTROL_PLANE_SDK."""
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_errors.py:8]: why the docstring sys this is an Exception for SDK ?

File: src/promptflow/promptflow/core/_errors.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    def __init__(
+        self,
+        message="",
+        message_format="",
+        target: ErrorTarget = ErrorTarget.CONTROL_PLANE_SDK,
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_errors.py:14]: Shall we add new ErrorTarget?

File: src/promptflow/promptflow/core/_flow.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+        :param params_override: Parameters to override, defaults to None
+        :type params_override: typing.Optional[list]
+        :return: Class to use for deserializing the data & its "type". Type will be None if no override is provided.
+        :rtype: tuple[class, typing.Optional[str]]
+        """
+        return cls, "flow"
+
+
+class Flow(FlowBase):
+    """This class is used to hide internal interfaces from user.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:130]: docstring need to update

File: src/promptflow/promptflow/core/_flow.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        from promptflow.core._flow_context_resolver import FlowContextResolver
+
+        invoker = FlowContextResolver.resolve(flow=self)
+        result = invoker._invoke(
+            data=inputs,
+        )
+        return result
+
+
+class EagerFlow(FlowBase):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:242]: When will this class be used?

File: src/promptflow/promptflow/core/_errors.py (hunk 5 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_errors.py:8]: let me change this.

File: src/promptflow/promptflow/core/_errors.py (hunk 6 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_errors.py:14]: OK

File: src/promptflow/promptflow/core/_flow.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+            else:
+                return Flow._load(path=flow_path, dag=data, content_hash=content_hash, **kwargs)
+
+    @classmethod
+    def load(
+        cls,
+        source: Union[str, PathLike],
+        raise_error=True,
+        **kwargs,
+    ):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:179]: add docstring and return annotation for public methods

Comment (wangchao1230) [reply src/promptflow/promptflow/core/_flow.py:179]: -> Flow

File: src/promptflow/promptflow/core/_flow.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        meta_dict = generate_flow_meta(
+            flow_directory=self.code,
+            source_path=self.entry_file,
+            entry=self.entry,
+            dump=False,
+        )
+        return ExecutableEagerFlow.deserialize(meta_dict)
+
+
+class AsyncFlow(Flow):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:304]: should have a load function?

File: src/promptflow/promptflow/core/_flow.py (hunk 8 above)
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:242]: can EagerFlow enherits Flow

File: src/promptflow/promptflow/core/_flow.py (hunk 10 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:304]: same with _load, it just override the Flow.load()
Flow.load could separate the Flows
EagerFlow.load could not.

File: src/promptflow/promptflow/core/_flow.py (hunk 8 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:242]: Yes, let's do it.

File: src/promptflow/promptflow/core/_flow.py [hunk 11]
```diff
@@ -127,13 +127,11 @@ def _resolve_cls_and_type(cls, data, params_override):
 
 
 class Flow(FlowBase):
-    """This class is used to hide internal interfaces from user.
+    """This class is public Flow interface.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:130]: refine this line to sth like below:

A flow is ...

File: src/promptflow/promptflow/core/_flow.py [hunk 12]
```diff
@@ -127,13 +127,11 @@ def _resolve_cls_and_type(cls, data, params_override):
...
 
-    User interface should be carefully designed to avoid breaking changes, while developers may need to change internal
-    interfaces to improve the code quality. On the other hand, making all internal interfaces private will make it
-    strange to use them everywhere inside this package.
-
-    Ideally, developers should always initialize ProtectedFlow object instead of Flow object.
+    Provides two basic functionalities:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:132]: May put a python example here

File: src/promptflow/promptflow/core/_flow.py [hunk 13]
```diff
@@ -171,31 +169,49 @@ def _dispatch_flow_creation(
...
+        **kwargs,
+    ) -> "Flow":
+        """
+        Load flow from YAML file.
+
+        :param source: The local yaml source of a flow. Must be a path to a local file.
+            If the source is a path, it will be open and read.
+            An exception is raised if the file does not exist.
+        :type source: Union[PathLike, str]
+        :param is_async_call: Optional argument to indicate the return value is an async function.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:197]: no need to provide this here

Comment (wangchao1230) [reply src/promptflow/promptflow/core/_flow.py:197]: User should do AsyncFlow.load()

File: src/promptflow/promptflow/core/_flow.py [hunk 14]
```diff
@@ -270,6 +289,53 @@ def _load(cls, path: Path, data: dict, **kwargs):
...
+            flow_directory=self.code,
+            source_path=self.entry_file,
+            entry=self.entry,
+            dump=False,
+        )
+        return ExecutableEagerFlow.deserialize(meta_dict)
+
+    def __call__(self, *args, **kwargs):
+        """Direct call of eager flow WILL cause exceptions."""
+        raise UserErrorException("Eager flow can not be called as a function.")
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:335]: Let's use the type name, not write "eager flow" in error message. which is not user facing

File: src/promptflow/promptflow/core/_flow.py [hunk 15]
```diff
@@ -239,8 +255,11 @@ def invoke(self, inputs: dict) -> "LineResult":
         return result
 
 
-class EagerFlow(FlowBase):
-    """This class is used to represent an eager flow."""
+class EagerFlow(Flow):
+    """This class is used to represent an eager flow.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:259]: Let's say non-dag flow here

File: src/promptflow/promptflow/core/_flow.py [hunk 16]
```diff
@@ -259,7 +278,7 @@ def __init__(
         # entry file name
         self.entry_file = self._resolve_entry_file(entry=entry, working_dir=code)
         # TODO(2910062): support eager flow execution cache
-        super().__init__(data=data, path=path, code=code, content_hash=None, **kwargs)
+        super().__init__(code=code, path=path, dag=data, content_hash=None, **kwargs)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:281]: why this change？
why it has dag parameter here?

File: src/promptflow/promptflow/core/_flow.py [hunk 17]
```diff
@@ -239,8 +255,11 @@ def invoke(self, inputs: dict) -> "LineResult":
         return result
 
 
-class EagerFlow(FlowBase):
-    """This class is used to represent an eager flow."""
+class EagerFlow(Flow):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:258]: Let's name the class as FlexFlow

File: src/promptflow/promptflow/core/_flow.py (hunk 12 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:132]: OK

File: src/promptflow/promptflow/core/_flow.py (hunk 11 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:130]: https://microsoft.github.io/promptflow/concepts/concept-flows.html ?

File: src/promptflow/promptflow/core/_flow.py (hunk 15 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:259]: Okay, let me change it thoroughly

File: src/promptflow/promptflow/core/_flow.py (hunk 16 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:281]: We have change this FlexFlow base class against Flow.

So the super() now is no longer FlowBase but Flow.

Flow init contains the paramter dag.
image.png

Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:281]: image

File: src/promptflow/promptflow/core/_flow.py [hunk 18]
```diff
@@ -127,11 +127,21 @@ def _resolve_cls_and_type(cls, data, params_override):
...
-
-    Provides two basic functionalities:
-    1. Load flow from YAML file: flow = Flow.load(source=src, ...)
-    2. Invoke loaded flow with inputs. flow() or flow.invoke(inputs=...)
+    """A Flow in the context of PromptFlow is a sequence of steps that define a task.
+    Each step in the flow could be a prompt that is sent to a language model, or simply a function task,
+    and the output of one step can be used as the input to the next.
+    Flows can be used to build complex applications with language models.
+
+    The Flow class provides two basic functionalities:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:135]: remove this part

File: src/promptflow/promptflow/core/_flow.py [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+        return cls, "flow"
+
+
+class Flow(FlowBase):
+    """A Flow in the context of PromptFlow is a sequence of steps that define a task.
+    Each step in the flow could be a prompt that is sent to a language model, or simply a function task,
+    and the output of one step can be used as the input to the next.
+    Flows can be used to build complex applications with language models.
+
+    Simple Example:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/core/_flow.py:135]: copy this: 
image

https://microsoft.github.io/promptflow/reference/python-library-reference/promptflow.tracing.html#promptflow.tracing.trace

Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:135]: image

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 20]
```diff
@@ -247,49 +38,56 @@ def __init__(
...
+        return self._flow_dir / self._dag_file_name
 
     @property
     def tools_meta_path(self) -> Path:
         target_path = self._flow_dir / PROMPT_FLOW_DIR_NAME / FLOW_TOOLS_JSON
         target_path.parent.mkdir(parents=True, exist_ok=True)
         return target_path
 
+    @property
+    def language(self) -> str:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:61]: Let's expose this class in future PR and refine the public filed here

Comment (crazygao) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:61]: ok，let me mark this.

**Review by chw-microsoft — APPROVED** [review]

File: src/promptflow/promptflow/core/_flow.py [hunk 21]
```diff
@@ -0,0 +1,423 @@
...
+        flow = Flow.load(source="path/to/flow.dag.yaml")
+        result = flow(input_a=1, input_b=2)
+
+    """
+
+    def __init__(
+        self,
+        code: Union[str, PathLike],
+        path: Union[str, PathLike],
+        dag: dict,
```
Comment (D-W-) [inline src/promptflow/promptflow/core/_flow.py:149]: Not sure if we want to expose this

File: src/promptflow/promptflow/core/_flow.py [hunk 22]
```diff
@@ -0,0 +1,423 @@
...
+        from promptflow.core._flow_context_resolver import FlowContextResolver
+
+        invoker = FlowContextResolver.resolve(flow=self)
+        result = invoker._invoke(
+            data=inputs,
+        )
+        return result
+
+
+class FlexFlow(Flow):
```
Comment (D-W-) [inline src/promptflow/promptflow/core/_flow.py:257]: Why inherit Flow instead of FlowBase?

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 23]
```diff
@@ -247,49 +38,56 @@ def __init__(
...
+        return self._data.get(LANGUAGE_KEY, FlowLanguage.Python)
+
+    @property
+    def additional_includes(self) -> list:
```
Comment (0mza987) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:65]: This `additional includes` gets defined  both in `Flow` and `EagerFlow`, can we move it to base class?

Comment (0mza987) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:65]: as well as `languange`

File: src/promptflow/promptflow/core/_flow.py (hunk 22 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:257]: Create a task for this issue.

File: src/promptflow/promptflow/core/_flow.py (hunk 21 above)
Comment (crazygao) [reply src/promptflow/promptflow/core/_flow.py:149]: let us change this in following items
